velox/connectors/clp/search_lib/ClpVectorLoader.cpp

```diff
@@ -14,16 +14,111 @@
  * limitations under the License.
  */
 
+#include <cmath>
+#include <cstdint>
 #include <utility>
 
 #include "clp_s/ColumnReader.hpp"
+#include "clp_s/SchemaTree.hpp"
 
 #include "velox/connectors/clp/search_lib/ClpVectorLoader.h"
+#include "velox/type/Timestamp.h"
 #include "velox/vector/ComplexVector.h"
 #include "velox/vector/FlatVector.h"
 
 namespace facebook::velox::connector::clp::search_lib {
 
+namespace {
+enum class TimestampPrecision : uint8_t {
+  Seconds,
+  Milliseconds,
+  Microseconds,
+  Nanoseconds
+};
+
+/**
+ * Estimates the precision of an epoch timestamp as seconds, milliseconds,
+ * microseconds, or nanoseconds.
+ *
+ * This heuristic relies on the fact that 1 year of epoch nanoseconds is
+ * approximately 1000 years of epoch microseconds and so on. This heuristic
+ * can be unreliable for timestamps sufficiently close to the epoch, but should
+ * otherwise be accurate for the next 1000 years.
+ *
+ * Note: Future versions of the clp-s archive format will adopt a
+ * nanosecond-precision integer timestamp format (as opposed to the current
+ * format which allows other precisions), at which point we can remove this
+ * heuristic.
+ *
+ * @param timestamp
+ * @return the estimated timestamp precision
+ */
+template <typename T>
+auto estimatePrecision(T timestamp) -> TimestampPrecision {
+  constexpr int64_t kEpochMilliseconds1971{31536000000};
+  constexpr int64_t kEpochMicroseconds1971{31536000000000};
+  constexpr int64_t kEpochNanoseconds1971{31536000000000000};
+  if (timestamp > kEpochNanoseconds1971) {
+    return TimestampPrecision::Nanoseconds;
+  } else if (timestamp > kEpochMicroseconds1971) {
+    return TimestampPrecision::Microseconds;
+  } else if (timestamp > kEpochMilliseconds1971) {
+    return TimestampPrecision::Milliseconds;
+  } else if (timestamp > -kEpochMilliseconds1971) {
+    return TimestampPrecision::Seconds;
+  } else if (timestamp > -kEpochMicroseconds1971) {
+    return TimestampPrecision::Milliseconds;
+  } else if (timestamp > -kEpochNanoseconds1971) {
+    return TimestampPrecision::Microseconds;
+  } else {
+    return TimestampPrecision::Nanoseconds;
+  }
```

Comment:
```suggestion
  auto absTimestamp = timestamp >= 0 ? timestamp : -timestamp;

  if (absTimestamp > kEpochNanoseconds1971) {
    return TimestampPrecision::Nanoseconds;
  } else if (absTimestamp > kEpochMicroseconds1971) {
    return TimestampPrecision::Microseconds;
  } else if (absTimestamp > kEpochMilliseconds1971) {
    return TimestampPrecision::Milliseconds;
  } else {
    return TimestampPrecision::Seconds;
  }
```
I think this is more clear. I also have a question for the cornercase, if timestamp == kEpochNanoseconds1971, the precision in the current implemention would be microseconds, but if timestamp == -kEpochNanoseconds1971, the precision would be nanoseconds. Is it intend to be different? (If so, please ignore my code suggestion).

Reply:
Your implementation is more readable -- let's use it.